akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/BehaviorBuilder.scala

```diff
@@ -34,7 +32,29 @@ final class BehaviorBuilder[T] private (messageHandlers: List[Case[T, T]], signa
   /**
    * Build a Behavior from the current state of the builder
    */
-  def build(): Behavior[T] = new BuiltBehavior(messageHandlers.reverse, signalHandlers.reverse)
+  def build(): Behavior[T] = {
+    if (shouldUseMapBasedBehavior(messageHandlers)) {
+      val classifiedMessageHandlers: Map[Class[_ <: T], Array[Case[T, T]]] =
+        messageHandlers.groupBy(_.`type`.get).map { case (clazz, list) => (clazz, list.toArray) }.toMap
+      new BuiltClassifiedBehavior[T](classifiedMessageHandlers, signalHandlers.reverse.toArray)
+    } else {
+      new BuiltBehavior[T](messageHandlers.reverse.toArray, signalHandlers.reverse.toArray)
+    }
+  }
+
+  private def shouldUseMapBasedBehavior(handlers: List[Case[T, T]]): Boolean = {
+    if (handlers.forall(_.`type`.isDefined) && messageHandlers.size >= 5) {
+      val classes = handlers.map(_.`type`.get)
+      val set = DisjointSet(classes)
+      for (x <- classes;
+           y <- classes if x != y && x.isAssignableFrom(y)) {
+        set.union(x, y)
+      }
+      classes.forall(clazz => clazz == set.find(clazz))
```

Comment:
require every class is not connected with others

Reply:
I don't quite see why we need the custom data type for this, since we anyway need to go through each unique pair of classes couldn't it just be something like 
```scala
val notSafe = classes.exists(c1 => classes.exists(c2 => (c1 != c2) && c1.isAssignableFrom(c2)))
``` 

(turning `classes` to set first to speed it up ofc)

Reply:
I think this can be later used to log some runtime warnning?as `String` case is been matched by `Object` case etcs?

Reply:
Maybe that's a IDE's work?

Reply:
If we want to add such warnings that could also be done inside a nested exists logic, but I don't think we should invest time in that for now. I think we should skip the DisjointSet unless we really have a reason to add such a datatype (and I don't see any such reason right now).

Reply:
Cool, let's discuss that latter。